content/posts/gsoc-2022-x86-il/index.md

```diff
@@ -0,0 +1,125 @@
+
+## Operands
+
+Also, there had to be a general interface for accessing the operands, irrespective of whether they were a register, or a memory location, or an immediate value. Writing the helper functions is easy, but providing ease of usage and removing duplication code requires thought. I have also slightly touched upon this in the [SuperH IL lifting post](https://dmaroo.github.io/superhlifting) on [my blog](https://dmaroo.github.io/).
+
+## IL implementation
+
+The descriptions in the x86 manual for some instructions get pretty complex and intense. Making sure that they are implemented correctly is not easy. This is exactly why we have tracetesting to verify the semantics, but manual verification never hurts. Also, the IL gets pretty large for many instructions, and to avoid this, it is important to choose a semantically correct, yet concise implementation. This involves removing unnecessary casting, using loops, removing duplicating and so on. In fact, removing duplication by using variables brings down the IL code size by a lot.
+
+## QEMU
+
+QEMU's codebase is a great for learning purposes. It consists of complex C constructs (tons of non-trivial preprocessor macros), but at the same time it is quite well-written, and also reasonably well documented. However, debugging the codebase is not that easy, since the code generates a buffer (`code_gen_buffer`), which then executes more code.
+
+# Future Work
+
+The work I have done in my duration is just the core part of the lifting. The lifting is far from being stable and ready-to-use. It is more of an pre-alpha release than a finished feature. The next steps will be as follows.
+
+ - **Tracetest**: Thoroughly tracetest the implementations to make sure the semantics are right. 
+ - **API**: Add Rizin commands and API, which would be wrappers around the IL, so as to visualize and access the IL. A similar graphical widget should also be added to Cutter.
+ - **Documentation**: Document the IL and its usage in the Rizin book.
+ - **Analysis**: Integrate the IL in the analysis loop to lead to a better analysis.
+ - **Instructions**: Add lifting for more x86 instructions.
+
```

Comment:
While not your future work, I recommend also mentioning the upcoming conversion of all architectures from ESIL to RZIL and extending RZIL with floats:
- https://github.com/rizinorg/rizin/issues/2080
- https://github.com/rizinorg/rizin/issues/1461

Reply:
Sure.